File: Backend/serverPython/controller/user_controller.py
from flask import request
import hashlib
from serverPython.database.db import get_db_conn
import json
from serverPython.s3.s3 import upload_file

def index():
    return {"status":200,"mensaje":"okay en python"}

def login():
    # user, password
    data = request.json
    user = data.get('user')
    password = hashlib.md5(data.get('password').encode()).hexdigest()
    print(user, password)

    # query = f"call sp_usuario(@mensaje, NULL, '{user}', NULL, '{password}', NULL, 'L');"
    
    db_conn = get_db_conn()
    try:
        cursor = db_conn.cursor()
        cursor.execute(query, multi=True)
    except Exception as error: 
        print()
    
    try:
        
        
        
        # obtener user id
        query = f"SELECT id_usuario FROM usuario WHERE usuario = '{user}' AND clave = '{password}';"
        # print("print ",data)
        cursor.execute(query)
        
        result_sets = cursor.fetchone()
        id_usuario = result_sets[0]
        print(id_usuario)

        cursor.close()
        db_conn.close()

        if id_usuario:
            return {"status":200,"mensaje":'Datos de autenticacion correctos.', "id_usuario":id_usuario}
        return {"status":401,"mensaje":"Credenciales incorrectos."}
    except Exception as error: 
        print(error)
        return {"status":400,"mensaje":"Datos de autenticacion incorrectos"}

# CALL sp_usuario(@mensaje, id, 'Usuario', 'Nombre', 'Password', 'URL', 'C');
def register():
    # user, name, password, url
    data = request.json
    user = data.get('user')
    name = data.get('name')
    password = hashlib.md5(data.get('password').encode()).hexdigest()
    url = data.get('url') # foto recibida

    nombre_foto = "FotoDePerfil1"
    # con el nombre de usuario, asignar nombre a la foto
    url_foto = "https://practica2-semi1-b-2s2024-imageness-g4.s3.amazonaws.com/" + "Fotos_Perfil/" + user + nombre_foto + ".jpg"

    # Subir al bucket S3
    # upload_file(url, nombre_foto, user)

    print(user, name, password, url)
    query = f"call sp_usuario(@mensaje, NULL, '{user}', '{name}', '{password}', '{url_foto}', 'C');"
    print(query)

    db_conn = get_db_conn()
    try:
        cursor = db_conn.cursor()
        cursor.execute(query)
        cursor.close()
        db_conn.close()
        return {"status":200,"mensaje":'Usuario registrado correctamente.'}
    except Exception as error: 
        print(error)
        return {"status":400,"mensaje":"error"}
    
def getProfile(id):
    # user, password
    print("id: ", id)
    #query = f"call sp_usuario(@mensaje, {id}, NULL, NULL, NULL, NULL, 'R');"
    query = f"SELECT id_usuario, usuario, nombre_completo, clave, url_foto FROM usuario WHERE id_usuario = {id};"
    print(query)

    db_conn = get_db_conn()
    try:
        cursor = db_conn.cursor()
        cursor.execute(query)

        # Fetch all result sets returned by the stored procedure
        result_sets = cursor.fetchall()
        
        cursor.close()
        db_conn.close()

         # Convert result_sets to a list of dictionaries
        data = [{"id_usuario": row[0], "usuario": row[1], "nombre_completo": row[2], "clave": row[3], "url_foto": row[4]} for row in result_sets]

        # Convert data to JSON format
        json_data = json.dumps(data[0], indent=4)

        return json_data
    except Exception as error: 
        print(error)
        return {"status": 400, "mensaje": "Datos no encontrados"}

def editProfile(id):
    data = request.json
    user = data.get('user')
    name = data.get('name')
    password = hashlib.md5(data.get('password').encode()).hexdigest()
    url = data.get('url')

    # parametrizado porque por algun motivo solo asi funciono aqui PERO EN OTROS NO FUNCIONO >:c
    query = "CALL sp_usuario(@mensaje, %s, %s, %s, %s, %s, 'U')"
    query_params = (id, user, name, password, url)

    db_conn = get_db_conn()
    try:
        cursor = db_conn.cursor()
        cursor.execute(query, query_params)
        cursor.close()
        db_conn.close()
        return {"status": 200, "mensaje": 'Datos actualizados.'}
    except Exception as error: 
        print(error)
        return {"status": 400, "mensaje": "Datos no actualizados"}

File: Backend/server/controller/user.controller.js
const query = require("../database/query");
const md5 = require("md5");
const s3 = require("../s3/s3");

const index = async (req, res) => {
  res.status(200).json({ message: "OK en index controller" });
};

// Login
// call practica1.sp_usuario(@mensaje, id_usuario, usuario, nombre_completo, password, url_foto, 'L');
// call practica1.sp_usuario(@mensaje, NULL, 'Usuario1', NULL, 'PWD', NULL, 'L');
const login = async (req, res) => {
  try {
    const { user, password } = req.body;

    // Validar que user y password están presentes
    if (!user || !password) {
      return res.status(400).json({
        status: 400,
        mensaje: "Usuario y contraseña son requeridos.",
      });
    }

    const pass = md5(password);
    console.log("password: " + password);
    console.log("pass: " + pass);

    // Ejecutar la consulta
    const [outcome] = await query.execute_(
      "call sp_usuario(@mensaje, NULL, ?, NULL, ?, NULL, 'L');",
      [user, pass]
    );

    // Manejo de errores en la respuesta
    if (outcome.err) {
      return res.status(400).json(outcome.err);
    } else {
      if (outcome[0][0]["p_id_usuario"]) {
        return res.status(200).json({
          status: 200,
          mensaje: "Credenciales correctas.",
          id_usuario: outcome[0][0]["p_id_usuario"],
        });
      } else {
        return res.status(401).json({
          status: 401,
          mensaje: "Credenciales incorrectas.",
        });
      }
    }
  } catch (error) {
    // Manejo de errores inesperados
    console.error("Error en el login: ", error);
    return res.status(500).json({
      status: 500,
      mensaje: "Error interno del servidor.",
    });
  }
};

// Registro
// CALL sp_usuario(@id, @mensaje, 'Usuario', 'Nombre', 'Password', 'URL', 'C');
const register = async (req, res) => {
  const { user, name, password, url } = req.body;
  nombre_foto = "FotoDePerfil1";

  let url_foto =
    "https://practica2-g4-imagenes.s3.us-east-2.amazonaws.com/" +
    "fotos/" +
    user +
    nombre_foto +
    ".jpg";

  // BUCKET S3
  //   s3.uploadPicture(url, nombre_foto, user);

  const pass = md5(password);
  console.log("pass " + pass);
  const [outcome] = await query.execute_(
    "call sp_usuario(@mensaje, NULL, ?, ?, ?, ?, 'C'); SELECT @mensaje;",
    [user, name, pass, url_foto]
  );
  if (outcome.err) {
    return res.status(400).json(outcome.err);
  } else {
    //return res.status(200).json(outcome[1][0]);
    return res
      .status(200)
      .json({ status: 200, mensaje: outcome[1][0]["@mensaje"] });
  }
};

// Obtener Perfil
// ID obtenido en el URL
// CALL sp_usuario(@id, @mensaje, 'Usuario', 'Nombre', 'Password', 'URL', 'R');
const getProfile = async (req, res) => {
  var user_id = req.params["id"];
  console.log("user id: " + user_id);
  let user = "user";
  let name = "name";
  let pass = "passs";
  let url = "default";
  await query.execute_("set @id = ?;", [user_id]);
  const outcome = await query.execute_(
    "call sp_usuario(@mensaje, @id, ?, ?, ?, ?, 'R');",
    [user, name, pass, url]
  );
  if (outcome.err) {
    return res.status(400).json(outcome.err);
  } else {
    return res.status(200).json(outcome[0][0][0]);
  }
};

// Editar Perfil
// ID obtenido en el URL
// CALL sp_usuario(@mensaje, id, 'Usuario', 'Nombre', 'Password', 'URL', 'U');
const editProfile = async (req, res) => {
  var user_id = req.params["id"];
  const { user, name, password, url } = req.body;
  const pass = md5(password);

  //verificar si la contrase;a es correcta
  const [outcomePassword] = await query.execute_(
    "call sp_usuario(@mensaje, NULL, ?, NULL, ?, NULL, 'L');",
    [user, pass]
  );
  if (!outcomePassword[0][0]["p_id_usuario"]) {
    return res.status(401).json({
      status: 401,
      mensaje: "Contraseña incorrecta. No se han actualizado los datos.",
    });
  }

  const outcome = await query.execute_(
    "call sp_usuario(@mensaje, ?, ?, ?, ?, ?, 'U'); SELECT @mensaje;",
    [user_id, user, name, pass, url]
  );

  if (outcome.err) {
    return res.status(400).json({ status: 400, mensaje: outcome.err });
  } else {
    return res
      .status(200)
      .json({ status: 200, mensaje: outcome[0][1][0]["@mensaje"] });
  }
};

module.exports = {
  index,
  login,
  register,
  getProfile,
  editProfile,
};
